src/Jackalope/Version/VersionHandler.php

```diff
@@ -174,4 +209,40 @@ public function checkoutItem($path)
 
         $this->session->save();
     }
+
+    /**
+     * Attaches a nt:frozenNode copy of $node to $versionNode
+     *
+     * @param NodeInterface $versionNode The version to attach the nt:frozenNode to
+     * @param NodeInterface $node The node to be copied into an nt:frozenNode
+     */
+    private function createFrozenNode(NodeInterface $versionNode, NodeInterface $node)
+    {
+        $frozenNode = $versionNode->addNode('jcr:frozenNode', 'nt:frozenNode');
+        $frozenNode->setProperty('jcr:frozenUuid', $node->getProperty('jcr:uuid'));
+        $frozenNode->setProperty('jcr:frozenPrimaryType', $node->getProperty('jcr:primaryType'));
+        if ($frozenNode->hasProperty('jcr:mixinTypes')) {
+            $frozenNode->setProperty('jcr:frozenMixinTypes', $node->getProperty('jcr:mixinTypes'));
+        }
+
+        foreach ($node->getProperties() as $property) {
+            /** @var PropertyInterface $property */
+            $propertyName = $property->getName();
+            if ($propertyName == 'jcr:primaryType'
+                || $propertyName == 'jcr:mixinTypes'
+                || $propertyName == 'jcr:uuid'
+            ) {
+                continue;
+            }
+
+            $onParentValue = $property->getDefinition()->getOnParentVersion();
+            if ($onParentValue != OnParentVersionAction::COPY && $onParentValue != OnParentVersionAction::VERSION) {
+                continue;
+            }
+
+            // TODO apply other steps based on onParentValue
```

Comment:
It's for later. Actually it looks to me like there should be done more according to the specification, but the phpcr-api-tests do not cover these cases. I thought I'll leave the todos here, so that later the phpcr-api-tests can be adapted accordingly.